server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -384,47 +389,75 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+                                final String shardPath =
+                                    shardContainer(shardResult.indexId, shardResult.shardId).path().buildAsString();
+                                return shardResult.blobsToDelete.stream().map(blob -> shardPath + blob);
+                            }),
+                            deleteResults.stream().map(shardResult -> shardResult.indexId).distinct().map(indexId ->
+                                indexContainer(indexId).path().buildAsString() + globalMetaDataFormat.blobName(snapshotId.getUUID()))
+                        ).map(absolutePath -> {
+                            assert absolutePath.startsWith(basePath);
+                            return absolutePath.substring(basePathLen);
+                        }).collect(Collectors.toList()));
+                },
+                // Any exceptions after we have updated the root level RepositoryData are only logged but won't fail the delete request
+                e -> logger.warn(
+                    () -> new ParameterizedMessage("[{}] Failed to delete some blobs during snapshot delete", snapshotId), e)),
+            () -> afterCleanupsListener.onResponse(null));
+
+        // Consumer to invoke once an updated ShardGeneration instance is ready to be written
+        final CheckedConsumer<ShardGenerations, IOException> onShardGenerations = shardGens -> {
+            final RepositoryData updatedRepoData = repositoryData.removeSnapshot(snapshotId, shardGens);
+            // Write out new RepositoryData
+            writeIndexGen(updatedRepoData, repositoryStateId, version);
+            // Run unreferenced blobs cleanup in parallel to snapshot deletion
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(afterCleanupsListener,
+                l -> cleanupStaleBlobs(foundIndices, rootBlobs, updatedRepoData, ActionListener.map(l, ignored -> null))));
+        };
+
+        if (version.onOrAfter(SnapshotsService.SHARD_GEN_IN_REPO_DATA_VERSION)) {
+            // New path that updates the pointer to each deleted shard's generation in the root RepositoryData
+
+            // Once we are done removing the snapshot from the metadata if each affected shard we can update the repository
+            // metadata atomically in a single step as follows:
+            // 1. Remove the snapshot from the list of existing snapshots
+            // 2. Update the index shard generations of all updated shard folders
+            //
+            // Note: If we fail updating any of the individual shard paths, none of them are changed since the newly created
+            //       index-${gen_uuid} will not be referenced by the existing RepositoryData and new RepositoryData is only
+            //       written if all shard paths have been successfully updated.
+            deleteIndices(repositoryData, snapshotId, version, ActionListener.wrap(res -> {
```

Comment:
the `deleteIndices` method name is confusing. That method does a lot of things.

Reply:
Tried to name it better and added a comment on what this actually does, see if it's a little clearer now.

In fact, I'd like to just flatten the whole delete method into one (I had that in a previous version of this PR) to make the order of operations easier to follow. Just didn't want to do it here to keep the diff smaller, but I can if it helps ... seems to me it would make things nicer.

Reply:
The diff on these methods is already very complex (I checked out the code in my IDE to make sense of it), so perhaps just give it a go.

Reply:
Flattened out the delete logic some more now and tried to add more comments and improve naming a little. Let me know if that helps.